ouroboros-consensus/src/Ouroboros/Consensus/NodeNetwork.hs

```diff
@@ -49,142 +61,237 @@ import           Ouroboros.Consensus.ChainSyncServer
--------------------------------------------------------------------------------}
-
--- TODO something nicer than eCS, eBF, bytesCS, and bytesBF. Mux them over one
--- channel.
-data NetworkProvides m up blk = NetworkProvides {
-      -- | Notify network layer of new upstream node
-      --
-      -- NOTE: Eventually it will be the responsibility of the network layer
-      -- itself to register and deregister peers.
-      addUpstream   :: forall eCS eBF bytesCS bytesBF.
-                       (Exception eCS, Exception eBF)
-                    => up
-                    -> NodeComms m (ChainSync (Header blk) (Point blk)) eCS bytesCS
-                       -- Communication for the Chain Sync protocol
-                    -> NodeComms m (BlockFetch blk)                     eBF bytesBF
-                       -- Communication for the Block Fetch protocol
-                    -> m ()
-
-      -- | Notify network layer of a new downstream node
-      --
-      -- NOTE: Eventually it will be the responsibility of the network layer
-      -- itself to register and deregister peers.p
-    , addDownstream :: forall eCS eBF bytesCS bytesBF.
-                       (Exception eCS, Exception eBF)
-                    => NodeComms m (ChainSync (Header blk) (Point blk)) eCS bytesCS
-                       -- Communication for the Chain Sync protocol
-                    -> NodeComms m (BlockFetch blk)                     eBF bytesBF
-                       -- Communication for the Block Fetch protocol
-                    -> m ()
-    }
+data SharedState m s a = SharedState (m s) (s -> a)
 
+instance Functor m => Functor (SharedState m s) where
+    fmap f (SharedState ms g) = SharedState ms (f . g)
 
--- | Required by the network layer to initiate comms to a new node
-data NodeComms m ps e bytes = NodeComms {
-      -- | Codec used for the protocol
-      ncCodec    :: Codec ps e m bytes
+-- A left biased semigroup instance.
+--
+-- TODO: needing this instance is another sign that we should either fix
+-- consensus client applications or generalise mux.  It is lawful, but it
+-- brings the arbitrary choice of being left biased.
+--
+instance Semigroup a => Semigroup (SharedState m s a) where
+    SharedState ms f <> SharedState _ g = SharedState ms (f <> g)
 
-      -- | Construct a channel to the node
-      --
-      -- This is in CPS style to allow for resource allocation.
-    , ncWithChan :: forall a. (Channel m bytes -> m a) -> m a
-    }
+runSharedState :: Monad m => SharedState m s a -> m a
+runSharedState (SharedState ms f) = ms >>= pure . f
+
+
+-- | Protocol handlers as provided by the consensus.
+--
+data ProtocolHandlers m peer blk = ProtocolHandlers {
+      phChainSyncClient
+        :: peer
+        -> ChainSyncClient (Header blk) (Point blk) m Void
+
+    , phChainSyncServer
+        :: ChainSyncServer (Header blk) (Point blk) m ()
 
+    -- TODO block fetch client does not have GADT view of the handlers.
+    , phBlockFetchClient
```

Comment:
Can you use the `BlockFetchClient` type synonym from https://github.com/input-output-hk/ouroboros-network/blob/2529bf78a5f8c6900fc74338473cea9785515eb7/ouroboros-network/src/Ouroboros/Network/BlockFetch/Client.hs#L58